### ISSUE TITLE
The chart is flapping in live stream chart.

### ISSUE THREAD

**Issue by future11** [amcharts/amcharts4#3549]
Hi,

I encountered the issue that the chart is flapping when I zoomed it fully.
I know the below line caused the issue. But I am not sure if this is a bug or wrong settings. I need your help.

if (date.getTime() > dateAxis.max - 2000)
      dateAxis.max = date.getTime() + 5000;

Please check the following code.

Thanks for your service.
Ben

// Themes begin
am4core.useTheme(am4themes_animated);
// Themes end

var chart = am4core.create("chartdiv", am4charts.XYChart);
chart.hiddenState.properties.opacity = 0;

chart.padding(0, 0, 0, 0);

chart.zoomOutButton.disabled = true;

var data = [];
var visits = 10;
var i = 0;

for (i = 0; i <= 30; i++) {
    visits -= Math.round((Math.random() < 0.5 ? 1 : -1) * Math.random() * 10);
    data.push({ date: new Date().setSeconds(i - 30), value: visits });
}

chart.data = data;

var dateAxis = chart.xAxes.push(new am4charts.DateAxis());
dateAxis.renderer.grid.template.location = 0;
dateAxis.renderer.minGridDistance = 30;
dateAxis.dateFormats.setKey("second", "ss");
dateAxis.periodChangeDateFormats.setKey("second", "[bold]h:mm a");
dateAxis.periodChangeDateFormats.setKey("minute", "[bold]h:mm a");
dateAxis.periodChangeDateFormats.setKey("hour", "[bold]h:mm a");
dateAxis.renderer.inside = true;
dateAxis.renderer.axisFills.template.disabled = true;
dateAxis.renderer.ticks.template.disabled = true;

var valueAxis = chart.yAxes.push(new am4charts.ValueAxis());
valueAxis.tooltip.disabled = true;
valueAxis.interpolationDuration = 500;
valueAxis.rangeChangeDuration = 500;
valueAxis.renderer.inside = true;
valueAxis.renderer.minLabelPosition = 0.05;
valueAxis.renderer.maxLabelPosition = 0.95;
valueAxis.renderer.axisFills.template.disabled = true;
valueAxis.renderer.ticks.template.disabled = true;

var series = chart.series.push(new am4charts.LineSeries());
series.dataFields.dateX = "date";
series.dataFields.valueY = "value";
series.interpolationDuration = 500;
series.defaultState.transitionDuration = 0;
series.tensionX = 0.8;

dateAxis.interpolationDuration = 500;
dateAxis.rangeChangeDuration = 500;

document.addEventListener("visibilitychange", function() {
    if (document.hidden) {
        if (interval) {
            clearInterval(interval);
        }
    }
    else {
        startInterval();
    }
}, false);

// add data
var interval;
function startInterval() {
    interval = setInterval(function() {
        visits =
            visits + Math.round((Math.random() < 0.5 ? 1 : -1) * Math.random() * 5);
        var lastdataItem = series.dataItems.getIndex(series.dataItems.length - 1);
        chart.addData(
            { date: new Date(lastdataItem.dateX.getTime() + 1000), value: visits },
            0
        );
    }, 1000);
}

startInterval();

// all the below is optional, makes some fancy effects
// gradient fill of the series
series.fillOpacity = 1;
var gradient = new am4core.LinearGradient();
gradient.addColor(chart.colors.getIndex(0), 0.2);
gradient.addColor(chart.colors.getIndex(0), 0);
series.fill = gradient;

// this makes date axis labels to fade out
dateAxis.renderer.labels.template.adapter.add("fillOpacity", function (fillOpacity, target) {
    var dataItem = target.dataItem;
    return dataItem.position;
})

// need to set this, otherwise fillOpacity is not changed and not set
dateAxis.events.on("validated", function () {
    am4core.iter.each(dateAxis.renderer.labels.iterator(), function (label) {
        label.fillOpacity = label.fillOpacity;
    })
})

// bullet at the front of the line
var bullet = series.createChild(am4charts.CircleBullet);
bullet.circle.radius = 5;
bullet.fillOpacity = 1;
bullet.fill = chart.colors.getIndex(0);
bullet.isMeasured = false;

series.events.on("validated", function() {
    bullet.moveTo(series.dataItems.last.point);
    bullet.validatePosition();
});

let xAxisRange;
chart.events.on("datavalidated", function () {
    if (!xAxisRange) {
        xAxisRange = dateAxis.axisRanges.create();
        xAxisRange.grid.stroke = am4core.color("red");
        xAxisRange.grid.strokeWidth = 1;
        xAxisRange.grid.strokeOpacity = 1;
        xAxisRange.label.disabled = true;
        xAxisRange.grid.strokeDasharray = "3,3";
    }

    let date = new Date(series.dataItems.last.dateX.getTime());
    date.setMilliseconds(dateAxis.baseDuration / 2);
    xAxisRange.date = date;

    // The following line caused the flapping issue.  
    if (date.getTime() > dateAxis.max - 2000)
      dateAxis.max = date.getTime() + 5000;
});

chart.cursor = new am4charts.XYCursor();
chart.cursor.behavior = "panX";

function zoomIn() {
    if (dateAxis.start <= 0.9)
      dateAxis.start += 0.1;
}

function zoomOut() {
    if (dateAxis.start >= 0.1)
      dateAxis.start -= 0.1;
}

**Comment by martynasma** [comment]
Can you please explain what you mean by "flapping"?

Your chart looks OK to me:
https://codepen.io/team/amcharts/pen/6fbce67063a61f70e402a932856bef8a?editors=0010

**Comment by future11** [comment]
Please add following two lines to html and change chartdiv height to 500px.
<button onclick="zoomIn()">Zoom In</button>
<button onclick="zoomOut()">Zoom Out</button>

And then if you click "Zoom In" button 9 times, you will see the issue.

**Comment by future11** [comment]
Sorry, I mean these two buttons
<button onclick="zoomIn()">Zoom In</button>
<button onclick="zoomOut()">Zoom Out</button>

**Comment by future11** [comment]
&lt;button onclick="zoomIn()"&gt;Zoom In&lt;/button&gt;
&lt;button onclick="zoomOut()"&gt;Zoom Out&lt;/button&gt;

**Comment by martynasma** [comment]
You may need to disable range animation for the ValueAxis:

```JavaScript
valueAxis.rangeChangeDuration = 0;
```

**Comment by future11** [comment]
I tried it but unfortunately, still same issue.

**Comment by martynasma** [comment]
I suppose you may need to restrict zooming in to a minimum number of time intervals, e.g.:

```JavaScript
dateAxis.minZoomCount = 10;
```

**Comment by future11** [comment]
Still same. But I fixed it by setting strictMinMax of dateAxis and valueAxis.
I have another question.
When I changed the chart data dynamically, how can I set pre-zooming?
In the doc, it can call zoom API inside "ready" event of the chart. But when I changed the chart data, for example, when clicking any button, this event is not called.
So I called zoom API in "datavalidated" event but some zoom related values are messing.
